### DISCUSSION TITLE
Good "Sample Plugins"?

### DISCUSSION THREAD

**Discussion by sernaferna** [mdx-editor/editor#569] [Q&A]
I need to build some **plugins** (I think), and I'm more than happy to look at the code, but I'm having trouble figuring out which plugins would be the best examples to copy from. 

I've got a couple of use cases; if someone has suggestions for plugins I could look at to get me started, I'd really appreciate it:

* **Simple ones**; e.g., `==text==` for highlighting or `^text^` for superscript.
  * I know there's a toolbar plugin that let's the user do superscript/subscript, but in my domain I'll be using superscript often enough that I prefer the more compact `^text^` notation than `<sup>` tags sprinkled through the markdown
  * If a plugin isn't necessary for this kind of simple use case I'm open to simpler solutions as well, if there's an example I can parrot
* **More complex ones**, such as markdown like `[|text (field2)|field3;flags]` that would render into a link. I haven't thought it through, but I'm guessing something like `<a href=field2 some-attr=field3>text</a>` kind of thing. I'll need something to differentiate it from other links so the export to markdown works -- I'll figure that part out. 😉
* **Bonus**: if there's a plugin or other code sample that will help me figure out how to add keyboard shortcuts, even better. 🙂

This feels like a basic question so I apologize for that, it's just that the plugins I've been looking at seem very heavily reliant on working with markdown notations Lexical already knows (e.g., bold formatting using `**`, or links using `[]()`, or quotes using `> `), but I don't see examples where the existing plugins are creating **new notations** in markdown, if that makes sense.

I've done similar things for **marked**, so I'm not afraid to get in there and write some regular expressions if required. 😉

**Comment by petyosi** [comment]
From your examples:

1) and 2) can be broken down to:

- An mdast extension that parses the you refer to as an AST node and that AST node back into markdown. Such extension should not directly embedded in the editor, because you will most likely need it when rendering the markdown for display purposes as well. You can find such extensions here https://github.com/syntax-tree/mdast?tab=readme-ov-file. 
- A Lexical code (usually a node, if such a node does not exist) that's the editor representation of the AST node you have parsed. Most existing built-in plugins are basically that, you can check their source code for further reference.
- The above code is glued together with visitors.
- Optionally, a toolbar item or any kind of additional "chrome" that you might need.

Most of what I've said above is [described in this help article](https://mdxeditor.dev/editor/docs/extending-the-editor#markdown--editor-state-conversion).

About shortcuts: this is 100% Lexical - the framework has a nice utility for capturing shortcuts.
